File: tests/fixtures/code/sample.ts
// A small search client, invented for the test suite.
import { Compendium } from "./compendium";

/** One ranked result. */
export interface Result {
  parentId: string;
  score: number;
}

export type Ranker = (results: Result[]) => Result[];

export enum Mode {
  Keyword,
  Semantic,
}

// Ranks results by score, highest first.
export function rank(results: Result[]): Result[] {
  return results.sort((left, right) => right.score - left.score);
}

export class Client {
  private compendium: Compendium;

  constructor(compendium: Compendium) {
    this.compendium = compendium;
  }

  // Runs one query and returns the ranked results.
  search(term: string, mode: Mode = Mode.Keyword): Result[] {
    return rank(this.compendium.lookup(term, mode));
  }
}

File: tests/fixtures/code/Sample.cs
// A small reader for a built index, invented for the test suite.
using System;
using System.Collections.Generic;

namespace Example.Search
{
    /// <summary>One ranked result.</summary>
    public interface IResult
    {
        string ParentId { get; }
    }

    // Reads a compendium and answers questions about it.
    public class Compendium
    {
        private readonly List<string> titles;

        public Compendium(List<string> titles)
        {
            this.titles = titles;
        }

        // Returns every title holding the term.
        public IEnumerable<string> Search(string term)
        {
            return titles.FindAll(title => title.Contains(term));
        }
    }
}
